File: ts/algos/sorting/MergeSort.ts
/**
 * Complexity
 *
 * Best - n
 * Average - n^2
 * Worst - n^2
 *
 * Memory - 1
 * Stable - Yes
 *
 *  */

export default function sort(
  arr: Array<number | string>
): Array<number | string> {
  if (arr.length <= 1) return arr;

  const middleIndex = Math.floor(arr.length / 2);
  const leftArray = arr.slice(0, middleIndex);
  const rightArray = arr.slice(middleIndex, arr.length);

  const leftSortedArray = sort(leftArray);
  const rightSortedArray = sort(rightArray);

  return merge(leftSortedArray, rightSortedArray);
}

function merge(
  arr1: Array<number | string>,
  arr2: Array<number | string>
): Array<number | string> {
  const sortedArray: Array<number | string> = [];

  let leftIndex = 0,
    rightIndex = 0;

  while (leftIndex < arr1.length && rightIndex < arr2.length) {
    let minElement = null;

    if (arr1[leftIndex] < arr2[rightIndex]) {
      minElement = arr1[leftIndex];
      leftIndex++;
    } else {
      minElement = arr2[rightIndex];
      rightIndex++;
    }

    sortedArray.push(minElement);
  }
  return sortedArray
    .concat(arr1.slice(leftIndex))
    .concat(arr2.slice(rightIndex));
}

console.log(sort(["d", "b", "a"]));
console.log(sort([2, 4, 21, 1, 44]));

File: python/algos/sorting/MergeSort.py
def merge(arr, p, q, r):
    left = arr[p : q + 1]
    right = arr[q + 1 : r + 1]

    i = j = 0
    k = p

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1
        k += 1

    while i < len(left):
        arr[k] = left[i]
        i += 1
        k += 1

    while j < len(right):
        arr[k] = right[j]
        j += 1
        k += 1


def merge_sort(arr, p, r):
    if p >= r:
        return
    q = (p + r) // 2
    merge_sort(arr, p, q)
    merge_sort(arr, q + 1, r)
    merge(arr, p, q, r)


def sort(arr):
    merge_sort(arr, 0, len(arr) - 1)


if __name__ == "__main__":
    arr = [2, 3, 4, 1, 6, 7, 3]
    print(arr, end=" ==> ")
    sort(arr)
    print(arr)
